### PR TITLE
Replace immutable TraceState with a mutable struct

### PR DESCRIPTION
**Pull request by bogdandrutu**
Fixes https://github.com/open-telemetry/opentelemetry-collector/issues/6022

Main motivation is that we may want to allow mutating (adding an entry to the state).

Signed-off-by: Bogdan <bogdandrutu@gmail.com>

### CODE REVIEW

File: pdata/ptrace/json_test.go [hunk 1]
```diff
@@ -84,7 +84,7 @@ var tracesOTLPFull = func() Traces {
 	sp.SetDroppedLinksCount(1)
 	sp.SetEndTimestamp(pcommon.NewTimestampFromTime(time.Now()))
 	sp.SetParentSpanID(spanID)
-	sp.SetTraceState("state")
+	sp.TraceStateStruct().FromRaw("state")
```
Comment (jpkrohling) [inline pdata/ptrace/json_test.go:87]: Can't we keep having the setter like it was before?

File: pdata/ptrace/traces.go [hunk 2]
```diff
@@ -70,14 +70,34 @@ func (ms Traces) ResourceSpans() ResourceSpansSlice {
 	return newResourceSpansSlice(&ms.getOrig().ResourceSpans)
 }
 
-// TraceState is a string representing the tracestate in w3c-trace-context format: https://www.w3.org/TR/trace-context/#tracestate-header
+// Deprecated: [v0.60.0] use pcommon.TraceState.
 type TraceState string
 
 const (
-	// TraceStateEmpty represents the empty TraceState.
+	// Deprecated: [v0.60.0] use pcommon.TraceState.
```
Comment (jpkrohling) [inline pdata/ptrace/traces.go:77]: People using this are using a specific instance of the trace state. Perhaps the godoc tell people how to accomplish the same with the new API?

File: pdata/ptrace/generated_traces.go [hunk 3]
```diff
@@ -620,14 +620,9 @@ func (ms Span) SetSpanID(v pcommon.SpanID) {
...
-// TraceState returns the tracestate associated with this Span.
-func (ms Span) TraceState() TraceState {
-	return TraceState(ms.getOrig().TraceState)
-}
-
-// SetTraceState replaces the tracestate associated with this Span.
-func (ms Span) SetTraceState(v TraceState) {
-	ms.getOrig().TraceState = string(v)
+// TraceStateStruct returns the tracestatestruct associated with this Span.
+func (ms Span) TraceStateStruct() pcommon.TraceState {
```
Comment (jpkrohling) [inline pdata/ptrace/generated_traces.go:624]: This name is going to change in the future, right?

**Review by jpkrohling — COMMENTED** [review]
I also prefer this over #6053

File: pdata/ptrace/json_test.go (hunk 1 above)
Comment (bogdandrutu) [reply pdata/ptrace/json_test.go:87]: The rule of pdata: If it is mutable we don't have a setter. If it is immutable then we have a setter. In this proposal the struct is mutable, so we don't offer a setter (users should change in place).

File: pdata/ptrace/generated_traces.go (hunk 3 above)
Comment (bogdandrutu) [reply pdata/ptrace/generated_traces.go:624]: Correct will become `TraceState`, but to do deprecation cannot reuse the same name.

File: pdata/ptrace/traces.go (hunk 2 above)
Comment (bogdandrutu) [reply pdata/ptrace/traces.go]: Clarified how to use it temporary. There are only 2 usages of this in contrib, will take as an action item to provide an API or a helper to check for empty.

File: CHANGELOG.md [hunk 4]
```diff
@@ -95,6 +95,11 @@
   - `SummaryDataPoint.Flags` -> `SummaryDataPoint.FlagsImmutable`
   - `MetricDataPointFlags` -> `MetricDataPointFlagsImmutable`
   - `NewMetricDataPointFlags` -> `MetricDataPointFlagsImmutable`
+- Deprecate `ptrace.TraceState` in favor of `pcommon.TraceState`. (#6052)
+  - `ptrace.Span.TraceState` in favor of `ptrace.Span.TraceStateStruct().AsRaw()`
+  - `ptrace.Span.SetTraceState` in favor of `ptrace.Span.TraceStateStruct().FromRaw`
+  - `ptrace.SpanLink.TraceState` in favor of `ptrace.SpanLink.TraceStateStruct().AsRaw()`
+  - `ptrace.SpanLink.SetTraceState` in favor of `ptrace.SpanLink.TraceStateStruct().FromRaw`
```
Comment (dmitryax) [inline CHANGELOG.md:102]: Let's maybe clarify here that `TraceStateStruct` is a temporary name that will be replaced back to `TraceState` in the next release?

**Review by dmitryax — APPROVED** [review]
One doc nit, otherwise LGTM

File: CHANGELOG.md (hunk 4 above)
Comment (bogdandrutu) [reply CHANGELOG.md:102]: Done.

**Review by dmitryax — APPROVED** [review]

**Review by codeboten — APPROVED** [review]
+1 to #6052 over #6053, i couldn't come up with a useful scenario where an immutable tracestate would be preferable

**Pull request merged**